### PR TITLE
CCPA Video Bug

### PR DESCRIPTION
**Pull request by SyntaxNode**
CCPA validation errors should result in warnings, not fatal errors. The video endpoint is missing the filter to ignore warnings. This PR also removes the invalid CCPA consent string from the request in further processing.

### CODE REVIEW

File: endpoints/openrtb2/amp_auction_test.go [hunk 1]
```diff
@@ -122,9 +122,8 @@ func TestAMPPageInfo(t *testing.T) {
 }
 
 func TestGDPRConsent(t *testing.T) {
-	consent := "BONV8oqONXwgmADACHENAO7pqzAAppY"
+	consent := "BOu5On0Ou5On0ADACHENAO7pqzAAppY"
```
Comment (SyntaxNode) [inline endpoints/openrtb2/amp_auction_test.go:125]: Unrelated. Noticed the poor test setup in looking at the adjacent CCPA tests.

File: endpoints/openrtb2/auction.go [hunk 2]
```diff
@@ -315,7 +315,12 @@ func (deps *endpointDeps) validateRequest(req *openrtb.BidRequest) []error {
 	}
 
 	if err := ccpaPolicy.Validate(); err != nil {
-		errL = append(errL, &errortypes.Warning{Message: fmt.Sprintf("CCPA value is invalid and will be ignored. (%s)", err.Error())})
+		errL = append(errL, &errortypes.InvalidPrivacyConsent{Message: fmt.Sprintf("CCPA consent is invalid and will be ignored. (%v)", err)})
+
+		ccpaPolicy.Value = ""
+		if err := ccpaPolicy.Write(req); err != nil {
+			errL = append(errL, fmt.Errorf("Unable to remove invalid CCPA consent from the request. (%v)", err))
+		}
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction.go:323]: We now remove the invalid value from the request. This prevents invalid CCPA values from being passed downstream to bidders.

File: endpoints/openrtb2/auction_test.go [hunk 3]
```diff
@@ -943,21 +943,23 @@ func TestCCPAInvalidValueWarning(t *testing.T) {
...
 	}
 
 	errL := deps.validateRequest(&req)
 
-	expectedError := errortypes.Warning{Message: "CCPA value is invalid and will be ignored. (request.regs.ext.us_privacy must contain 4 characters)"}
-	assert.ElementsMatch(t, errL, []error{&expectedError})
+	expectedWarning := errortypes.InvalidPrivacyConsent{Message: "CCPA consent is invalid and will be ignored. (request.regs.ext.us_privacy must contain 4 characters)"}
+	assert.ElementsMatch(t, errL, []error{&expectedWarning})
+
+	assert.Empty(t, req.Regs.Ext, "Invalid Consent Removed From Request")
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction_test.go:962]: New assertion to verify the invalid consent string is removed.

File: endpoints/openrtb2/sample-requests/video/video_invalid_sample.json [hunk 4]
```diff
@@ -1,68 +1,69 @@
 {
```
Comment (SyntaxNode) [inline endpoints/openrtb2/sample-requests/video/video_invalid_sample.json:1]: I normalized the format of the sample json files. I removed the no longer existing `accountid` field, moved the no longer valid `gdpr` section (it's uses the standard IAB locations now), and reformatted the `user ` section to match IAB spec from an earlier customized `user` object. .. oh, and ran it through a json "prettifier".

File: endpoints/openrtb2/sample-requests/video/video_invalid_sample.json [hunk 5]
```diff
@@ -1,68 +1,69 @@
...
+      "series": "TvName",
+      "season": "season3",
+      "len": 900,
+      "livestream": 0
     },
-    "yob": 1991,
-    "gender": "F",
-    "keywords": "Hotels, Travelling"
-  },
-  "device11": {
```
Comment (SyntaxNode) [inline endpoints/openrtb2/sample-requests/video/video_invalid_sample.json:25]: Was this a typo or an intention way to effectively remove the device info? I erred on the side of a typo and fixed it.

File: endpoints/openrtb2/video_auction.go [hunk 6]
```diff
@@ -203,8 +203,9 @@ func (deps *endpointDeps) VideoAuctionEndpoint(w http.ResponseWriter, r *http.Re
 	// Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
 	deps.setFieldsImplicitly(r, bidReq) // move after merge
 
+	// here be the problem!
 	errL = deps.validateRequest(bidReq)
-	if len(errL) > 0 {
+	if errortypes.ContainsFatalError(errL) {
```
Comment (SyntaxNode) [inline endpoints/openrtb2/video_auction.go:208]: This is the main fix. I added similar filters to AMP and Auction, but forgot about Video.

File: endpoints/openrtb2/video_auction_test.go [hunk 7]
```diff
@@ -860,12 +861,12 @@ func TestParseVideoRequestWithUserAgentAndHeader(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Failed to fetch a valid request: %v", err)
 	}
-
 	headers := http.Header{}
 	headers.Add("User-Agent", "TestHeader")
 
 	deps := mockDeps(t, ex)
-	req, valErr, podErr := deps.parseVideoRequest(reqData, headers)
+	reqBody := string(getRequestPayload(t, reqData))
```
Comment (SyntaxNode) [inline endpoints/openrtb2/video_auction_test.go:868]: This is part of normalizing the json file formats.

File: privacy/ccpa/policy.go [hunk 8]
```diff
@@ -32,6 +32,10 @@ func ReadPolicy(req *openrtb.BidRequest) (Policy, error) {
 // Write mutates an OpenRTB bid request with the context of the CCPA policy.
 func (p Policy) Write(req *openrtb.BidRequest) error {
 	if p.Value == "" {
+		return clearPolicy(req)
```
Comment (SyntaxNode) [inline privacy/ccpa/policy.go:35]: It used to do nothing, but now there's a use case to remove the value.

File: endpoints/openrtb2/video_auction_test.go [hunk 9]
```diff
@@ -1036,6 +1042,74 @@ func TestCreateImpressionTemplate(t *testing.T) {
...
+		response := &openrtb_ext.BidResponseVideo{}
+		if err := json.Unmarshal(responseBytes, response); err != nil {
+			t.Fatalf("%s: Unable to unmarshal response.", test.description)
+		}
+		assert.Len(t, ex.lastRequest.Imp, 11, test.description+":imps")
+		assert.Len(t, response.AdPods, 5, test.description+":adpods")
+	}
+}
+
+func TestCCPAInvalidConsent(t *testing.T) {
```
Comment (jmaynardxandr) [inline endpoints/openrtb2/video_auction_test.go:1110]: Remove? The invalid CCPA string is covered by the 3rd case in above function I believe.

Comment (SyntaxNode) [reply endpoints/openrtb2/video_auction_test.go]: Ooops! You're absolutely right. This is a useless empty test I forgot to remove when I turned the above one into a "table driven" format. Removed.

File: endpoints/openrtb2/video_auction.go [hunk 10]
```diff
@@ -232,7 +232,7 @@ func (deps *endpointDeps) VideoAuctionEndpoint(w http.ResponseWriter, r *http.Re
 	}
 
 	if acctIdErr := validateAccount(deps.cfg, labels.PubID); acctIdErr != nil {
-		errL = append(errL, acctIdErr)
+		errL := []error{err}
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply endpoints/openrtb2/video_auction.go:235]: I added that bit just now. Yes, it's intentional. The previous expectation from the code is for `errL` to be empty here. If it had an error before my changes, that should've triggered an early exit. The video endpoint doesn't have a structured response for errors (yet? hopefully?) so there's currently no way to communicate the warning back to the caller. Best to drop it than for it to randomly show up along with another unrelated error such as an invalid account.

**Pull request merged**